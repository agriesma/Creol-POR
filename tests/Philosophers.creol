interface Phil
begin
  op borrowStick
  op returnStick
end

interface Butler
begin
  with Phil op getNeighbor(out neighbor: Phil)
end

class Philosopher(butler: Butler) implements Phil 
begin 
  var hungry: Bool
  var chopstick: Bool
  var neighbor: Phil
  var history: String
 
  op init == history := ""; chopstick := true; hungry := false;
    butler.getNeighbor(; neighbor);
 
  op run == !think(); ||| !eat(); ||| !digest(); 

  op digest == await wait ; hungry := true; history := plus(history, "d");
    await wait; !digest();

  op eat == await hungry; l!neighbor.borrowStick(); await l?; await chopstick;
    history := plus(history, "e"); hungry := false; !neighbor.returnStick();
    await wait; !eat();

  op think == await not hungry; history := plus(history, "t");
    await wait; !think();

  with Phil 
    op borrowStick == await chopstick; chopstick := false;
    op returnStick == chopstick := true;
 
end

class Butler implements Butler
begin
  var p1: Phil
  var p2: Phil
  var p3: Phil
  var p4: Phil
  var p5: Phil

  op init ==
    skip;

  op run ==
    p1 := new Philosopher(this);
    p2 := new Philosopher(this);
    p3 := new Philosopher(this);
    p4 := new Philosopher(this);
    p5 := new Philosopher(this);

  with Phil
    op getNeighbor(out neighbor: Phil) ==
      if caller = p1 then neighbor := p2; else
      if caller = p2 then neighbor := p3; else
      if caller = p3 then neighbor := p4; else
      if caller = p4 then neighbor := p5; else neighbor := p1; fi fi fi fi

end

