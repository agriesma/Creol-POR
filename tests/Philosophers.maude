load interpreter
mod PROGRAM is
pr INTERPRETER .
op init : -> Configuration [ctor] .
eq init =
< "Philosopher" : Cl | Inh: noInh, Par: "butler", Att: "butler" |-> null , "hungry" |-> null , "chopstick" |-> null , "neighbor" |-> null , "history" |-> null, Mtds: 
  < "init" : Mtdname | Param: noAid, Latt: noSubst, Code: "history" ::= str("") ; "chopstick" ::= bool(true) ; "hungry" ::= bool(false) ; "butler" . "getNeighbor" ( emp ; "neighbor" ) ; return ( emp ) > *
  < "run" : Mtdname | Param: noAid, Latt: noSubst, Code: ( "Dummy" ! "this" . "think" ( emp ) ||| "Dummy" ! "this" . "eat" ( emp ) ||| "Dummy" ! "this" . "digest" ( emp ) ) ; return ( emp ) > *
  < "digest" : Mtdname | Param: noAid, Latt: noSubst, Code: ( await wait ) ; "hungry" ::= bool(true) ; "history" ::= "plus" ( "history" # str("d") ) ; ( await wait ) ; "Dummy" ! "this" . "digest" ( emp ) ; return ( emp ) > *
  < "eat" : Mtdname | Param: noAid, Latt: noSubst, Code: ( await "hungry" ) ; "l" ! "neighbor" . "borrowStick" ( emp ) ; ( await ( "l" ?? ) ) ; ( await "chopstick" ) ; "history" ::= "plus" ( "history" # str("e") ) ; "hungry" ::= bool(false) ; "Dummy" ! "neighbor" . "returnStick" ( emp ) ; ( await wait ) ; "Dummy" ! "this" . "eat" ( emp ) ; return ( emp ) > *
  < "think" : Mtdname | Param: noAid, Latt: noSubst, Code: ( await "not" ( "hungry" ) ) ; "history" ::= "plus" ( "history" # str("t") ) ; ( await wait ) ; "Dummy" ! "this" . "think" ( emp ) ; return ( emp ) > *
  < "borrowStick" : Mtdname | Param: noAid, Latt: noSubst, Code: ( await "chopstick" ) ; "chopstick" ::= bool(false) ; return ( emp ) > *
  < "returnStick" : Mtdname | Param: noAid, Latt: noSubst, Code: "chopstick" ::= bool(true) ; return ( emp ) >, Ocnt: 0 >

< "Butler" : Cl | Inh: noInh, Par: noAid, Att: "p1" |-> null , "p2" |-> null , "p3" |-> null , "p4" |-> null , "p5" |-> null, Mtds: 
  < "init" : Mtdname | Param: noAid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noAid, Latt: noSubst, Code: "p1" ::= new "Philosopher" ( "this" ) ; "p2" ::= new "Philosopher" ( "this" ) ; "p3" ::= new "Philosopher" ( "this" ) ; "p4" ::= new "Philosopher" ( "this" ) ; "p5" ::= new "Philosopher" ( "this" ) ; return ( emp ) > *
  < "getNeighbor" : Mtdname | Param: noAid, Latt: "neighbor" |-> null, Code: if "equal" ( "caller" # "p1" ) th "neighbor" ::= "p2" el if "equal" ( "caller" # "p2" ) th "neighbor" ::= "p3" el if "equal" ( "caller" # "p3" ) th "neighbor" ::= "p4" el if "equal" ( "caller" # "p4" ) th "neighbor" ::= "p5" el "neighbor" ::= "p1" fi fi fi fi ; return ( "neighbor" ) >, Ocnt: 0 >

.
endm
