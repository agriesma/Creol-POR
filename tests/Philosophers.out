==========================================
reduce in PROGRAM : init .
result Configuration: 
< "Butler" : Cl | Inh: noInh, Par: noAid, Att: "p1" |-> null, "p2" |-> null,
    "p3" |-> null, "p4" |-> null, "p5" |-> null, Mtds:
  < "getNeighbor" : Mtdname | Param: noAid, Latt: "neighbor" |-> null, Code: if
    "="("caller" # "p1") th "neighbor" ::= "p2" el if "="("caller" # "p2") th
    "neighbor" ::= "p3" el if "="("caller" # "p3") th "neighbor" ::= "p4" el if
    "="("caller" # "p4") th "neighbor" ::= "p5" el "neighbor" ::= "p1" fi fi fi
    fi ; return("neighbor") > *
  < "init" : Mtdname | Param: noAid, Latt: noSubst, Code: skip ; return(emp) >
    *
  < "run" : Mtdname | Param: noAid, Latt: noSubst, Code: "p1" ::= new
    "Philosopher"("this") ; "p2" ::= new "Philosopher"("this") ; "p3" ::= new
    "Philosopher"("this") ; "p4" ::= new "Philosopher"("this") ; "p5" ::= new
    "Philosopher"("this") ; return(emp) >,
  Ocnt: 0 >

< "Philosopher" : Cl | Inh: noInh, Par: "butler", Att: "butler" |-> null,
    "chopstick" |-> null, "hist" |-> null, "hungry" |-> null, "neighbor" |->
    null, Mtds:
  < "borrowStick" : Mtdname | Param: noAid, Latt: noSubst, Code: await
    "chopstick" ; ("chopstick" assign bool(false)) ; return(emp) > *
  < "digest" : Mtdname | Param: noAid, Latt: noSubst, Code: release ; ("hungry"
    assign bool(true)) ; "hist" ::= "plus"("hist" # str("d")) ; release ;
    "anon.???" ! "this" . "digest"(emp) ; return(emp) > *
  < "eat" : Mtdname | Param: noAid, Latt: noSubst, Code: await "hungry" ; "l" !
    "neighbor" . "borrowStick"(emp) ; await ("l" ??) ; await "chopstick" ;
    "hist" ::= "plus"("hist" # str("e")) ; ("hungry" assign bool(false)) ;
    "anon.1" ! "neighbor" . "returnStick"(emp) ; release ; "anon.???" ! "this"
    . "eat"(emp) ; return(emp) > *
  < "init" : Mtdname | Param: noAid, Latt: noSubst, Code: ("hist" assign str(
    "")) ; ("chopstick" assign bool(true)) ; ("hungry" assign bool(false)) ;
    "sync.0" ! "butler" . "getNeighbor"(emp) ; "sync.0" ?("neighbor") ; return(
    emp) > *
  < "returnStick" : Mtdname | Param: noAid, Latt: noSubst, Code: ("chopstick"
    assign bool(true)) ; return(emp) > *
  < "run" : Mtdname | Param: noAid, Latt: noSubst, Code: ("anon.???" ! "this" .
    "digest"(emp) ||| "anon.???" ! "this" . "eat"(emp) ||| "anon.???" ! "this"
    . "think"(emp)) ; return(emp) > *
  < "think" : Mtdname | Param: noAid, Latt: noSubst, Code: await "~"("hungry")
    ; "hist" ::= "plus"("hist" # str("t")) ; release ; "anon.???" ! "this" .
    "think"(emp) ; return(emp) >,
  Ocnt: 0 >
Bye.
